Add quantization (dynamic)

avishreekh opened the pull request:
Solves #72

Het-Shah commented:
So we need to check whether this is working or not right? So benchmark it this way train a ResNet normally on MNIST or any dataset report its accuracy. Then do quantization and then report accuracy after quantization. Also, report the size of weights before and after quantization.

avishreekh commented:
> So we need to check whether this is working or not right? So benchmark it this way train a ResNet normally on MNIST or any dataset report its accuracy. Then do quantization and then report accuracy after quantization. Also, report the size of weights before and after quantization.

Sure

Het-Shah commented on logs.rst:
```diff
@@ -69,3 +69,15 @@ Results for Pruning Algorithms -
+Results for  Dynamic Quantization - 
+===================================
+
++------------+---------+---------------+----------------+-----------------+----------+--------------+
+| Model Used | Dataset |    Model      |    Size (MB)   | Training epochs | Accuracy | Time elapsed |
++============+=========+===============+================+=================+==========+==============+
+|            |         |     Base      |       4.15     |                 |  0.9582  |  28.84 s     |
+|            |         +---------------+----------------+                 +----------+--------------+
+|  ResNet152 |  MNIST  |   Quantized   |       0.56     |        1        |  0.9583  |  28.71 s     |
++------------+---------+---------------+----------------+-----------------+----------+--------------+
```
Can you explain the columns and the whole process that you followed once?

Het-Shah replied:
I don't understand the time elapsed column for example

avishreekh replied:
Oh sure.
Process: I trained a ResNet152 for one epoch on MNIST. After training it was quantized. The accuracy obtained by the non-quantized model was 0.9582 and it took 28.84 secs for it to run on the test set on CPU. The quantized model achieved an accuracy of 0.9583 and it took 28.71 secs for it to run on the test set on CPU.
Time elapsed is the time the model took to run on the test set.

Here is a link to a similar analysis by PyTorch:
https://pytorch.org/tutorials/advanced/dynamic_quantization_tutorial.html

Het-Shah replied:
How many epochs did you train ResNet for?

avishreekh replied:
One epoch

Het-Shah replied:
Can you perform a similar analysis for CIFAR-10 with more training epochs maybe?

avishreekh replied:
I wouldn't mind but why is that needed?

Het-Shah commented:
Now that we have figured out that this is working. We should probably add it to the Base KD class as well using `*args`.

avishreekh commented:
And what about the training on CIFAR?

Het-Shah commented:
> And what about the training on CIFAR?

Yes this too obviously. 

Additionally, we should add some helper functions to report results for quantization. For example, the model size calculation and inference time of non-quantized and quantized model.

avishreekh commented:
> > And what about the training on CIFAR?
> 
> Yes this too obviously.
> 
> Additionally, we should add some helper functions to report results for quantization. For example, the model size calculation and inference time of non-quantized and quantized model.

Should I add these to quantization class?

Het-Shah commented:
> Should I add these to quantization class?

Yes. Do you think anything else is needed?

avishreekh commented:
No I believe that should work

Het-Shah commented:
Hmm ok then

avishreekh commented:
> And what about the training on CIFAR?

Can I do this on pretrained ResNet from torchvision models?

avishreekh commented:
Check this and let me know if it looks good

Het-Shah commented on KD_Lib/Quantization/dynamic/dynamic_quantization.py:
```diff
@@ -9,14 +10,35 @@ class Dynamic_Quantizer:
+        print(f"Size of quantized_model (MB): {quantized_size}")
+
+
+### Helper Functions
+
+
+def get_size_of_model(model):
+    torch.save(model.state_dict(), "temp.p")
+    model_size = os.path.getsize("temp.p") / 1e6
+    os.remove("temp.p")
```
Move this to the class itself with `_get_size_of_model()` name coz its a helper function

avishreekh replied:
But that function is not essential to that class

Het-Shah reviewed (commented):
This seems good now.

avishreekh commented:
Ok so too many changes in a PR. Sorry for that.
Major features:
1. Added a quantization base class with functionalities for comparing sizes and inference performance.
2. Added static quantization module

Het-Shah commented on KD_Lib/Quantization/common/base_class.py:
```diff
@@ -0,0 +1,74 @@
+		"""
+        raise NotImplementedError
+
+    def compare_model_sizes(self):
+        """
```
a better name would be `get_model_sizes()` from users perspective

avishreekh replied:
I thought of the same previously. However, named it this way coz it will be used for comparison. I'll change it though, `get_model_sizes` seems nicer

Het-Shah commented on KD_Lib/Quantization/common/base_class.py:
```diff
@@ -0,0 +1,74 @@
+        print("-" * 80)
+        print(f"Size of original model (MB): {original_size}")
+        print(f"Size of quantized_model (MB): {quantized_size}")
+
+    def compare_inference_performance(self, data_loader):
+        """
```
Same here

Het-Shah commented on KD_Lib/Quantization/static/static_quantization.py:
```diff
@@ -0,0 +1,66 @@
+        qconfig=torch.quantization.default_qconfig,
+    ):
+        """
+		Function used for quantization
+        :param data_loader(torch.utils.data.DataLoader): DataLoader used for calibration
+        :param criterion(torch Loss_fn): Loss function used for calibration
+        :param num_calibration_batches(int): Number of batches used for calibration
+        :param qconfig: Configuration used for quantization
+		"""
```
Add an empty line between description and params, otherwise autodoc won't pick it up properly

Het-Shah commented on KD_Lib/Quantization/static/static_quantization.py:
```diff
@@ -0,0 +1,66 @@
+
+        return self.quantized_model
+
+    def _calibrate_model(self, data_loader, num_batches, criterion):
+        self.quantized_model.eval()
+        correct = 0
```
Add docstrings to helper functions as well or else they won't get picked up by autodoc

Het-Shah reviewed (commented):
Seems like great work to me. Seems ready just small stuff here and there.

Het-Shah commented on logs.rst:
```diff
@@ -69,3 +69,19 @@ Results for Pruning Algorithms -
+Results for  Quantization - 
+===================================
+
++------------+---------+---------------+----------------+-----------------+----------+--------------+
+| Model Used | Dataset |    Model      |    Size (MB)   | Training epochs | Accuracy | Time elapsed |
++============+=========+===============+================+=================+==========+==============+
+|  ResNet152 |  MNIST  |     Base      |       4.15     |        1        |  0.9582  |  28.84 s     |
+|            |         +---------------+----------------+                 +----------+--------------+
+|            |         |   Quantized   |       0.56     |                 |  0.9583  |  28.71 s     |
++            +---------+---------------+----------------+-----------------+----------+--------------+
+|            | CIFAR10 |     Base      |       4.15     |        1        |  0.9582  |  28.84 s     |
+|            |         +---------------+----------------+                 +----------+--------------+
+|            |         |   Quantized   |       0.56     |                 |  0.9583  |  28.71 s     |
++------------+---------+---------------+----------------+-----------------+----------+--------------+
```
One more thing here it will be good to mention which method did we use for quantization.

Het-Shah replied:
How are both MNIST and CIFAR10 getting the same metrics :P

avishreekh replied:
They are not lol. I just copy pasted to enter the result later but forgot

avishreekh commented:
Why is this build failing?

Het-Shah commented:
black

avishreekh commented:
I have done that
